src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -224,106 +250,127 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
         {
 #if DEBUG
             // Verify that loadProjectRootElement delegate does not call ProjectRootElementCache.Get().
-            s_getEntriesNumber++;
+            using var reentrancyGuard = new ReentrancyGuard();
+
+            // Verify that we never call this with _locker held, as that would create a lock ordering inversion with the per-file lock.
             ErrorUtilities.VerifyThrow(
-                s_getEntriesNumber == 1,
-                "Reentrance to the ProjectRootElementCache.Get function detected.");
+                !System.Threading.Monitor.IsEntered(_locker),
+                "Detected lock ordering inversion in ProjectRootElementCache.");
+#endif
+            // Should already have been canonicalized
+            ErrorUtilities.VerifyThrowInternalRooted(projectFile);
 
-            try
+            // First try getting the ProjectRootElement from the cache.
+            ProjectRootElement projectRootElement = GetOrLoad(projectFile, loadProjectRootElement: null, isExplicitlyLoaded, preserveFormatting);
+
+            if (projectRootElement != null || loadProjectRootElement == null)
             {
-#endif
-#pragma warning disable format // the release configuration fails as try-catch block is conditioned only for Debug; making the formatting in correct in Release.
-                // Should already have been canonicalized
-                ErrorUtilities.VerifyThrowInternalRooted(projectFile);
+                // If we found it or no load callback was specified, we are done.
+                return projectRootElement;
+            }
 
-                ProjectRootElement projectRootElement;
-                lock (_locker)
+            try
+            {
+                // We are about to load. Take a per-file lock to prevent multiple threads from duplicating the work multiple times.
+                object perFileLock = _fileLoadLocks.GetOrAdd(projectFile, () => new object());
+                lock (perFileLock)
                 {
-                    _weakCache.TryGetValue(projectFile, out projectRootElement);
+                    // Call GetOrLoad again, this time with the OpenProjectRootElement callback.
+                    return GetOrLoad(projectFile, loadProjectRootElement, isExplicitlyLoaded, preserveFormatting);
+                }
+            }
+            finally
+            {
+                // Remove the lock object as we have otherwise no good way of preventing _fileLoadLocks from growing unboundedly.
```

Comment:
I don't understand this argument. I believe multiple Get calls can happen in separate threads, right? The ReentrancyGuard only operates in DEBUG mode. If two start at about the same time, the _fileLoadLocks will prevent the second one from starting until the first is finished. Once the first finishes, the second starts (with the same object) while the first deletes it from the ConcurrentDictionary. If a third comes along while the second is still calling GetOrLoad, that third would make a new object and jump right in. This is unlikely to be relevant in practice, since it requires one thread to be on line 266-272 and a second thread to be sitting at line 276 when a third thread executes line 288, and then the first and second thread have to both be on line 279 for there to potentially be a problem, but it feels like an unnecessary risk to me.

Reply:
The rationale is that when a thread comes here and deletes the locker object from the dictionary, the project file must already be loaded (i.e. it was loaded by the first thread that took the lock, which would likely - but not necessarily - be the one doing the delete here). Therefore it is perfectly fine if other threads race to enter `GetOrLoad` simultaneously after this point, because it will be just a fast dictionary lookup, no actual loading will be taking place. In other words, we use the lock to allow only one thread to do the loading. Once loading is done, we don't care.

The alternative solutions would be:
- Never delete anything from `_fileLoadLocks`. The concern is that there may be pathological scenarios where the number of project files does not have a practical limit and the process would OOM.
- Hook into the `Scavenge` operation of `_weakCache`. That's where the weak cache is getting rid of entries for GC-collected items. Can be done with a callback but feels too complicated for no additional benefit.

Curious to hear your thoughts, thank you!

Reply:
Ah, so it sounds like it's ok for multiple threads to be inside the perFileLock as long as the cache has already been prepped. That seems reasonable. I'm trying to think through what the expected cases are, what the worst case is, and how likely that worst case is. It seems unlikely that there would be so many project files that the process would actually OOM specifically due to this lock cache, but there's also the constant incremental of having a larger dictionary to look through. Is a ConcurrentDictionary tree-based? If so, then if the average size of the ConcurrentDictionary with this implementation is 2^x, and this doubles the number of accesses, then from a time perspective, there'd have to be 2^2x projects for them to be even equally fast. From a memory, perspective, the pathological case is one in which a lot of threads manage to get past the first GetOrLoad call and stop before trying to access the lock. If they go in in pairs, there could theoretically be an extra threads/2 objects allocated. That also feels _very_ unlikely to me, though.

Overall, I think I'm convinced this is a good way to avoid doing more locking than the current version does. Do you happen to have numbers on how efficient locks are vs. using the ConcurrentDictionary?

I'm also starting to wonder if there's actually any reason to have a separate strong cache vs. weak cache, but I haven't looked into that difference too much yet. It feels like we're doing a lot of work to maintain a strong cache that we just use to make sure things from the weak cache don't get gc'd, which kinda defeats the point of using a weak cache...

Reply:
For the worst case scenarios where leaving the lock objects in the dictionary would be a problem, I was thinking more about cases where a long-lived MSBuild process would be asked to evaluate generated projects. I.e. something generates `%TEMP%\<guid>.proj`, asks MSBuild to evaluate/build it, then deletes it, all this in a loop. That could be problematic in practice, I think.

> Do you happen to have numbers on how efficient locks are vs. using the ConcurrentDictionary?

Reading upon the the `ConcurrentDictionary` implementation, it looks like it should be able to satisfy reads lock-free. In our use case where the dictionary is used more or less on the slow path, I don't think the choice of the data structure is critical. Using a concurrent collection nicely communicates the intent and an avoiding another explicit lock in our code means one less place to worry about potential deadlocks.

For the weak vs. strong cache it looks like the idea is to cache N recently loaded projects + whatever manages to survive GC. I agree that it would be useful to investigate how well this performs compared to alternative designs.

Reply:
That actually makes a lot of sense. I hadn't looked back to whether we clear the cache after every build, and I'd assumed we do, but it looks like we only do if the user requests it:
https://github.com/dotnet/msbuild/blob/60ea2f733cfcc78da965ab43cf040fe5c58800fb/src/Build/BackEnd/BuildManager/BuildManager.cs#L2117-L2124

I'm guessing we never actually ran that "experiment"; I'm wondering if we should?